File: cpp/triangle/triangle.cpp
#include "triangle.h"

namespace triangle {
	

	flavor kind(double a, double b, double c)
	{
		if (a <= 0 || b <= 0 || c <= 0) throw std::domain_error("Triangles cannot have negative side lengths");
		if (a + b < c || b + c < a || c + a < b) throw std::domain_error("Triangle is invalid");
		if (a == b && b == c) return flavor::equilateral;
		if (a == b || b == c || a == c) return flavor::isosceles;
		return flavor::scalene;
	}
}  // namespace triangle

File: python/triangle/triangle.py
def triangle(sides):
    return sum(sides) > max(sides) * 2

def equilateral(sides):
    return triangle(sides) and len(set(sides)) == 1

def isosceles(sides):
    return triangle(sides) and len(set(sides)) <= 2

def scalene(sides):
    return triangle(sides) and len(set(sides)) == 3
